util: fix parseEnv incorrectly splitting multiple ‘=‘ in value

rayark1 opened the pull request:
Previously, parseEnv would create multiple environment variables if a single line contained multiple ‘=‘ characters (e.g. A=B=C would become { A: ‘B=C’, B: ‘C’ }).
This commit ensures that only the first ‘=‘ is used as the key-value delimiter, and the rest of the line is treated as the value.

Fixes: https://github.com/nodejs/node/issues/57411

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

Armanidashh commented:
1

aduh95 commented on benchmark/fixtures/valid.env:
```diff
@@ -64,4 +65,4 @@ export EXPORT_EXAMPLE = ignore export
 MULTI_NOT_VALID_QUOTE="
 MULTI_NOT_VALID=THIS
-IS NOT MULTILINE
+IS NOT MULTILINE
```
Is this change related?

aduh95 commented on test/fixtures/dotenv/valid.env:
```diff
@@ -64,4 +65,4 @@ export EXPORT_EXAMPLE = ignore export
 MULTI_NOT_VALID_QUOTE="
 MULTI_NOT_VALID=THIS
-IS NOT MULTILINE
+IS NOT MULTILINE
```
Same here
```suggestion
IS NOT MULTILINE

```

jasnell commented on test/fixtures/dotenv/valid.env:
```diff
@@ -6,6 +6,7 @@ BASIC=basic
 
 # previous line intentionally left blank
 AFTER_LINE=after_line
+A=B=C
```
To me this just feels... odd. I would expect this is to be a formatting error unless the second `=` is escaped somehow. Like `A="B=C"` or `A=B\=C`, etc.

anonrig replied:
Why not escape it? I kind of agree that we should just throw maybe?

rayark1 replied:
Most libraries for handling .env files—whether it’s Python’s python-dotenv, Node.js’s dotenv, or others—treat only the first equal sign as the delimiter between the key and value. Everything following the first equal sign is considered part of the value.

Since this behavior mirrors that of POSIX shells, users naturally expect it.
<img width="484" alt="스크린샷 2025-03-14 오전 4 21 33" src="https://github.com/user-attachments/assets/976b134c-a1af-46f1-9fb2-2df39a330108" />

jasnell replied:
> In my opinion, one key advantage of using .env files is their simplicity. They work just like environment variables in POSIX shells, where only the first equal sign (=) separates the key from its value, and everything after is considered part of the value. For example, typing A=B=C in a POSIX shell assigns B=C to the variable A.

I'm not going to block on this I just find it a bit odd for a data format. 

Note that there are already inconsistencies between the env file and a regular environment variable. For example, in the env file I can do:

```
B=A B C
```

and the `process.env.B` will equal `'A B C'` without any escaping, but I cannot do `B=A B C ./node` and expect the same result.

I'm not going to block on it and the code changes in this PR LGTM but my general preference would be to instead require escaping of the second `=` in this case.

anonrig replied:
> I find the result... interesting.

Now I am interested as well but I'm not in front of the computer. What's the result?

lemire replied:
@anonrig Prior to this PR:

```
{ '"B': 'C"', A: 'B=C' }
```

It is hard to parse.

jasnell replied:
@lemire  .. there's no doubt the current parsing is buggy

```
// What we get currently
> parseEnv('A="B=C"')
{ '"B': 'C"', A: 'B=C' }
```

```
// What it should be
> parseEnv('A="B=C"')
{ A: 'B=C' }
```

So in that sense this PR is perfectly good., and that example has the second `=` escaped within quotes. My issue is with the unescaped variant. I personally feel (but won't block on) that `parseEnv('A=B=C')` should be an error.

lemire replied:
@jasnell 

> I personally feel (but won't block on)

I think your point is reasonable.

What seems undeniable is that the current behaviour is... surprising. :-)

anonrig commented on src/node_dotenv.cc:
```diff
@@ -145,7 +145,13 @@ void Dotenv::ParseContent(const std::string_view input) {
+      if (newline != std::string_view::npos) {
+        content.remove_prefix(newline + 1);
+      } else {
+        content.remove_prefix(content.size());
```
Nit

```suggestion
        content = {};
```

anonrig commented on src/node_dotenv.cc:
```diff
@@ -216,7 +224,7 @@ void Dotenv::ParseContent(const std::string_view input) {
           value = content.substr(0, newline);
           store_.insert_or_assign(std::string(key), value);
-          content.remove_prefix(newline);
+          content.remove_prefix(newline + 1);
```
Why do we make this change? Can you add a comment

rayark1 replied:
Previously, we used `content.remove_prefix(newline)`, which would leave the `'\n'` intact. On the next iteration, `content.front()` might be `'\n'`, potentially causing a re-check of the same line or treating it as an empty line.

anonrig commented on src/node_dotenv.cc:
```diff
@@ -226,8 +234,11 @@ void Dotenv::ParseContent(const std::string_view input) {
+        } else {
+          content.remove_prefix(content.size());
         }
+        continue;
```
Why do we continue here can you add a comment?

anonrig commented on src/node_dotenv.cc:
```diff
@@ -243,11 +254,21 @@ void Dotenv::ParseContent(const std::string_view input) {
+        if (hash_char != std::string_view::npos) {
+          value = content.substr(0, hash_char);
+        }
+        value = trim_spaces(value);
+
+        store_.insert_or_assign(std::string(key), value);
```
```suggestion
        store_.insert_or_assign(std::string(key), trim_spaces(value));
```

anonrig commented on src/node_dotenv.cc:
```diff
@@ -243,11 +254,21 @@ void Dotenv::ParseContent(const std::string_view input) {
+        value = trim_spaces(value);
+
+        store_.insert_or_assign(std::string(key), value);
+        content.remove_prefix(content.size());
```
```suggestion
        content = {};
```

anonrig commented on test/parallel/test-dotenv.js:
```diff
@@ -82,3 +82,4 @@ assert.strictEqual(process.env.DONT_EXPAND_SQUOTED, 'dontexpand\\nnewlines');
 assert.strictEqual(process.env.EXPORT_EXAMPLE, 'ignore export');
 // Ignore spaces before double quotes to avoid quoted strings as value
 assert.strictEqual(process.env.SPACE_BEFORE_DOUBLE_QUOTES, 'space before double quotes');
+assert.strictEqual(process.env.A, 'B=C');
```
you should add a test that validates a key with `"B` doesn't exist, so we don't regress.

'"B': 'C"'

anonrig commented on src/node_dotenv.cc:
```diff
@@ -145,7 +145,13 @@ void Dotenv::ParseContent(const std::string_view input) {
+      } else {
+        content.remove_prefix(content.size());
+      }
+      continue;
```
Can you add a comment to here explaining why we are continuing?

aduh95 commented:
Also, we should keep a test case for the unquoted version `A=B=C`, either to verify it throws, or whatever behavior that we can get consensus on.

rayark1 commented:
I have incorporated all the review comments and added the corresponding test cases. Please let me know if there are any further suggestions or improvements needed.

mcollina reviewed (approved):
lgtm
